Effects: Add support for GUIs and reference implementation for Audio Unit

fwcd opened the pull request:
### Based on #13887, fixes #13147

This PR adds a generic interface for effect backends for exposing GUIs in the form of `QDialog`s, along with a reference implementation for Audio Unit effects on macOS[^1]:

![image](https://github.com/user-attachments/assets/00a601ee-c7fd-40b8-a830-2dcfec303b8c)

Effect UIs are toggled via a new button, implemented in LateNight/Palemoon (the only skin for now) and shown for effects that provide a UI:

![image](https://github.com/user-attachments/assets/7df2f701-a629-4fac-b48f-19ac3aa79e6a)

![image](https://github.com/user-attachments/assets/0d75fbc7-a1f8-4ce5-ad8c-48a94f13eac8)

These can be hidden via the skin settings to reduce UI clutter, if not needed:

![image](https://github.com/user-attachments/assets/38ef3c0a-0725-49ac-af8e-fedbf62bdf9b)

### To do

- Generic implementation
  - [x] Add button for toggling effect UIs
  - [x] Add `DlgEffect` abstraction that sets up a generic effect UI (a floating dialog)
  - [x] Update the `ui_shown` control object when the window is closed manually
  - [x] Make sure that effect UIs are also closed upon closing the main window
  - [x] Figure out if we can gray out/hide the button for toggling the UI if the effect doesn't support UIs. `EffectManifest` now exposes a `hasUI` property, which may be helpful for that.
- Audio Unit implementation
  - [x] Host plugin views by loading them from the corresponding bundle
  - [x] Fall back to generic view if the plugin does not provide a custom UI
  - [x] Automatically size the dialog window correctly

### Potentially out-of-scope

- Fix sampling rate mismatch between Audio Unit plugins and Mixxx
  - See https://github.com/mixxxdj/mixxx/pull/13888#issuecomment-2477938796
- Update the parameter knobs on the Mixxx side after changing them from the GUI (the other direction already works and the effect also updates as intended from the GUI, this seems to be purely cosmetic)
- Sync tempo/beat with AU effects
  - Currently plugins that operate primarily in beat domain (such as ShaperBox) are not super useful, since they can't use the correct tempo (and their default tempo is almost always misaligned)
  - See e.g. https://forums.developer.apple.com/forums/thread/46085

Feedback welcome!

[^1]: Support for UIs of other effect backends like LV2 (or VST plugins once we implement #9019) could be added in future PRs as needed.

fwcd committed:
EffectManifest: Add hasUI

fwcd committed:
LateNight: Add button for toggling UI

fwcd committed:
Tooltips: Add effect UI button tooltip

fwcd committed:
LateNight: Make visibility of effect UI buttons configurable

fwcd committed:
EffectProcessor: Add createUI

fwcd committed:
AudioUnitManifest: Set hasUI using hasCustomView

fwcd committed:
EngineEffect: Add createUI

fwcd committed:
Skins: Use snake_case for ui_shown control

fwcd committed:
EffectSlot: Manage effect UI

fwcd committed:
AudioUnitEffectProcessor: Set up demo view

fwcd committed:
AudioUnitEffectProcessor: Load actual UI

fwcd committed:
AudioUnitManager: Expose name

fwcd committed:
AudioUnitEffectProcessor: Include name of audio unit in log messages

fwcd committed:
AudioUnitEffectProcessor: Improve error messages

fwcd committed:
AudioUnitEffectProcessor: Add missing nil check for cocoaViewInfo

fwcd committed:
AudioUnitEffectProcessor: Add missing initialization of cocoaViewInfo

fwcd committed:
AudioUnitEffectProcessor: Fall back to generic AU view if needed

fwcd commented:
I wonder why there a bunch of frequencies around 10k are missing according to spectrum analyzers in plugins. For comparison, here's some white noise in both Logic and Mixxx, viewed with the TDR Prism spectrum analyzer:

| Logic | Mixxx |
| - | - |
| <img src="https://github.com/user-attachments/assets/6e4f3a1c-0310-458c-ac2b-85944bdd7d6c" width="300"> | !<img src="https://github.com/user-attachments/assets/493a75de-a132-4ebb-bd33-d47153ca9cb3" width="360"> |

Notably, this does not show up when I open the recorded mix in Logic:

<img src="https://github.com/user-attachments/assets/c189fa82-bf26-4005-976b-b662e373ec5e" width="300">

I wonder if there's something wrong in the effect implementation (passing the audio buffer in some strange way) or whether there's something elsewhere in the Mixxx engine that could be causing this?

Swiftb0y commented on src/effects/backends/effectprocessor.h:
```diff
@@ -148,6 +155,20 @@ class EffectProcessorImpl : public EffectProcessor {
         return 0;
     }
 
+    std::unique_ptr<QDialog> createUI() override {
+        QLabel* label = new QLabel();
+        label->setText("This effect has no UI");
+
+        QVBoxLayout* layout = new QVBoxLayout();
+        layout->addWidget(label);
+
+        std::unique_ptr<QDialog> dialog = std::make_unique<QDialog>();
+        dialog->setWindowTitle("No UI available");
+        dialog->setLayout(layout);
+
+        return dialog;
+    }
```
is this really needed? This should only be shown when a plugin has `hasUI` is true even though it shouldn't be right?

fwcd replied:
Exactly, I haven't gotten around to do that check yet and decided that this was nicer than an empty dialog 😄 

I've also considered dropping the `hasUI` and just returning an optional dialog here, but that wouldn't give the UI a chance to hide/style the UI button beforehand.

Swiftb0y replied:
yeah... I think hiding the button if not available would be definitely good to have. I'd prefer to keep that if we can.

fwcd replied:
One consequence is that we keep a dialog open when the user switches to an effect that doesn't provide a UI. On the one hand I think this could be nice (e.g. the user might arrange their effect UIs on the screen in a particular way and it might be nice to keep them open until explicitly closed). On the other hand, the blanket implementation doesn't really do anything useful, so we might as well also just close the UI, in which case the next effect with a UI will automatically open the dialog since the control object enabling the UI is still toggled on.

Don't really have a preference what the best/most intuitive behavior is here. Any thoughts?

Swiftb0y replied:
I think we should follow the approach DAWs usually take. You can open as many effect UIs as there are effects and if you replace one effect with one that doesn't have a window or you remove the effect, the window just closes. Now you usually don't replace effects in DAWs directly (you insert a new one and remove the old one) but in mixxx we can replace an effect directly since we have slots. In that case we can consider replacing the old UI with the new UI directly (if the old effect had its UI open). Wdyt?

fwcd replied:
> Now you usually don't replace effects in DAWs directly

Depends on the DAW, at least in Logic the workflow is similar to Mixxx, where effects are essentially also slots with a dropdown.

> in mixxx we can replace an effect directly since we have slots. In that case we can consider replacing the old UI with the new UI directly (if the old effect had its UI open). Wdyt?

Sounds reasonable. The way the API is designed currently, we'd still instantiate a new dialog (the new effect UI might want to set its own title/resizability/etc.), but I was thinking of maybe just carrying over the position manually.

Swiftb0y replied:
> Depends on the DAW, at least in Logic the workflow is similar to Mixxx, where effects are essentially also slots with a dropdown.

ah yeah. My primary experience lies with bitwig and ableton... 

> but I was thinking of maybe just carrying over the position manually.

That may not be feasible on all platforms (not particularly relevant for MacOS I guess, but I don't think we could implement the same on wayland for example).

Swiftb0y commented on src/engine/effects/engineeffect.h:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <QDialog>
```
forward decl suffices actually

fwcd replied:
So just `class QDialog;`? Are forward decls generally preferred for headers?

Swiftb0y replied:
its a tradeoff IMO. If the header is big (usually the case with Qt headers) and the header where its included is used in lots of other headers and you only heap-allocate the class anyways (with a unique_ptr for example), then I'd forward-declare them. In this case it would make sense because you really only need to name it in order to return from a single member function. So pulling in an entire header does come with quite some bloat IMO.

Swiftb0y commented on src/effects/backends/audiounit/audiouniteffectprocessor.h:
```diff
@@ -70,5 +70,5 @@ class AudioUnitEffectProcessor final : public EffectProcessorImpl<AudioUnitEffec
     void syncStreamFormat(const mixxx::EngineParameters& engineParameters);
 
-    NSView* _Nullable createNativeUI(CGSize size);
+    NSView* _Nullable createNativeUI(CGSize size, QString& outError);
```
consider returning a `std::variant<NSView* _Nullable, QString>` (as in for example Rusts `std::Result` type). This would nicely map to `std::expected`.

fwcd replied:
I wasn't sure how well that meshes with the ObjC type, since there's reference counting and stuff involved, but I'll give that a try

Swiftb0y replied:
right... yeah idk enough about that to evaluate how feasible it is. I just assumed C++ semantics hold. If you see more value in a different approach then use that.

Swiftb0y reviewed (commented):
couple surface level comments. We should consider creating more C++like friendly wrappers for the underlying native C calls...

fwcd committed:
AudioUnitEffectProcessor: Set window style mask

fwcd committed:
AudioUnitEffectProcessor: Add missing NSWindowStyleMaskUtilityWindow

fwcd committed:
AudioUnitEffectProcessor: Automatically resize the dialog to fit

fwcd committed:
AudioUnitEffectProcessor: Factor out DlgAudioUnit

fwcd committed:
DlgAudioUnit: Disable resizing again

Since the dialog automatically adjusts its size based on the view,
manual resizing of the window should no longer be needed. The dominant
convention seems to be that audio plugins have a fixed size UI and some
plugins (such as TDR Prism) even provide their own resize control, we
will honor this by disabling direct window resizing.

fwcd committed:
DlgAudioUnit: Use std::variant for error handling

fwcd committed:
EngineEffect: Only forward-declare QDialog for effect UI

fwcd committed:
AudioUnitManifest: Set hasUI to true for all audio units

fwcd commented on src/effects/backends/audiounit/dlgaudiounit.mm:
```diff
@@ -0,0 +1,159 @@
+                             queue:[NSOperationQueue mainQueue]
+                        usingBlock:^(NSNotification* notification) {
+                            NSView* audioUnitView =
+                                    (NSView*)notification.object;
```
Small implementation note: We deliberately extract the view from the notification again to avoid creating a potential retain cycle by capturing the view again

fwcd committed:
EffectSlot: Automatically update UI when effect changes

fwcd committed:
Effects: Abstract out DlgEffect

fwcd committed:
DlgEffect: Add support for embedding custom UI

fwcd committed:
DlgAudioUnit: Embed native UI using the intended API

Instead of accessing the underlying NSView directly, which is
unsupported and fragile (e.g. changing it from Qt may just wipe it out),
we use the official `QWindow::fromWinId` and
`QWidget::createWindowContainer` to wrap native views. This also lets us
share/abstract out generic effect dialog logic, such as thinner
titlebars or floating-ness across other platforms or effect backends.

fwcd committed:
DlgEffect: Make it a floating tool window

This uses a smaller title bar on macOS and automatically causes the
dialog to float above other windows.

fwcd committed:
DlgEffect: Add transitive include

fwcd committed:
DlgAudioUnit: Set fixed size on dialog

fwcd committed:
DlgEffect: Add virtual destructor

fwcd committed:
DlgAudioUnit: Properly remove resize observer in destructor

fwcd committed:
DlgAudioUnit: Use non-fixed-size resize again (for now)

Unfortunately, this seems to work too unreliably for plugins that
implement their own draggable resize handle (apparently we don't get a
frame changed notification if the outer window is already fixed and the
plugin attempts to change the view frame?)

fwcd committed:
DlgEffect: Emit closed signal on close

fwcd committed:
DlgEffect: Add short doc comment

fwcd committed:
EffectSlot/DlgEffect: Synchronize close state properly

fwcd committed:
EffectSlot/LateNight: Show/hide effect UI button based on effect

fwcd committed:
LateNight: Move effect UI button to after the enable button

fwcd committed:
EffectSlot: Persist effect UI position

fwcd committed:
EffectProcessor: Remove blanket dialog title

fwcd committed:
EffectProcessor: Word doc comment on UI more accurately

fwcd committed:
LateNight: Remove accidental change

fwcd commented:
Marking this one as a draft until we figure out

- #13887

fwcd committed:
EffectSlot: Trigger updateEffectUI via effectChanged

fwcd committed:
EffectSlot: Update UI button in updateEffectUI

This handles the case where the user switches to the "empty" effect
correctly, which previously would keep the button visible.

fwcd committed:
DlgAudioUnit: Only apply resize observer to custom UI

fwcd committed:
DlgAudioUnit: Wait for instantiation before loading UI

This fixes a race condition

JoergAtGithub commented:
If you consider "Update the parameter knobs on the Mixxx side after changing them from the GUI" out of scope of this PR, than please create an issue with target `2.6-beta` for this!
Please consider also, what this means for controller mappings with soft-takeover for the parameter knobs.

JoergAtGithub commented on src/effects/backends/effectprocessor.h:
```diff
@@ -148,6 +155,19 @@ class EffectProcessorImpl : public EffectProcessor {
         return 0;
     }
 
+    std::unique_ptr<DlgEffect> createUI() override {
```
This seems wrong, as you declare this as pure virtual in this same file line 119: `virtual std::unique_ptr<DlgEffect> createUI() = 0;`

I would expect the GUI button disabled instead of showing a dialog that this effect has no GUI.

fwcd replied:
> This seems wrong, as you declare this as pure virtual in this same file line

The pure virtual signature applies to `EffectProcessor`, which serves as the interface, this method is in `EffectProcessorImpl`

> I would expect the GUI button disabled instead of showing a dialog that this effect has no GUI.

We do that too, but there are other advantages to keeping the dialog available, see this discussion for rationale: https://github.com/mixxxdj/mixxx/pull/13888#discussion_r1843617674
